### PR TITLE
[TACHYON-877] add LineageDoesNotExistException#check to improve readibility

### PR DESCRIPTION
**Pull request by miasantreble**
https://tachyon.atlassian.net/browse/TACHYON-877

@yupeng9 so I added one precondition to check that lineage we are looking for actually exists. Since we use the same error message template (`LINEAGE_DOES_NOT_EXIST`) in all but one case (`LINEAGE_NO_OUTPUT_FILE`), I made the former the default template but also provided a version where user can specify a different template. Let me know about your thoughts and I can continue to add more preconditions. I also noticed for some templates, we have similar templates in `PreconditionMessage` and `ExceptionMessage`, right now I'm sticking to `PreconditionMessage` as long as I could

### CODE REVIEW

**Comment by yupeng9** [comment]
@miasantreble , I think a `TachyonPrecondition`will make the code more readable. However, the proposed design introduces more code than it removes. Could we have a generic method `check` in `TachyonPrecondition`, and have a `Exception` type mapped to each exception message in `ExceptionMessage`, so that the precondition could instantiate the exception on the fly?

**Comment by miasantreble** [comment]
@yupeng9 thanks for your input, I think your design is less cumbersome since we don't need to create new `TachyonPrecondition` methods for each `TachyonException` I will update the PR accordingly

**Comment by yupeng9** [comment]
@miasantreble , any update on this PR?

**Comment by yupeng9** [comment]
After second-thought, with my proposed solution `check` can throw Tachyon `TachyonException `, which is not so desirable because it's too general.

**Comment by yupeng9** [comment]
Perhaps each exception can have its own check function? And we can move the function into `TachyonException` super class?

**Comment by yupeng9** [comment]
And add a static `check` method to each Exception class that calls the function. So it'll be like 
```
LineageDoesNotExistException.check(boolean expression, String errorMessage)
```
@miasantreble , what do you think?

**Comment by miasantreble** [comment]
@yupeng9 `check` function added to `LineageDoesNotExistException` in the latest upload. Let me know about your comments. Thanks!

**Comment by yupeng9** [comment]
Thanks @miasantreble, the new change looks good to me. Is this PR still WIP?

File: servers/src/main/java/tachyon/master/lineage/LineageMaster.java [hunk 1]
```diff
@@ -369,7 +367,8 @@ private void asyncCompleteFileFromEntry(AsyncCompleteFileEntry entry) {
    * @throws InvalidPathException if the file path is invalid
    */
   public synchronized LineageCommand lineageWorkerHeartbeat(long workerId,
-      List<Long> persistedFiles) throws FileDoesNotExistException, InvalidPathException {
+      List<Long> persistedFiles)
+          throws FileDoesNotExistException, InvalidPathException, LineageDoesNotExistException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/LineageMaster.java]: Can you update the javadoc?

File: common/src/main/java/tachyon/exception/LineageDoesNotExistException.java [hunk 2]
```diff
@@ -36,4 +36,11 @@ public LineageDoesNotExistException(ExceptionMessage message, Object... params)
   public LineageDoesNotExistException(ExceptionMessage message, Throwable cause, Object... params) {
     this(message.getMessage(params), cause);
   }
+
+  public static void check(boolean expression, ExceptionMessage message, Object... params)
```
Comment (yupeng9) [inline common/src/main/java/tachyon/exception/LineageDoesNotExistException.java]: Please add Javadoc for public method.

File: servers/src/main/java/tachyon/master/lineage/LineageMaster.java [hunk 3]
```diff
@@ -387,7 +386,7 @@ public synchronized LineageCommand lineageWorkerHeartbeat(long workerId,
   /**
    * @return the list of all the {@link LineageInfo}s
    */
-  public synchronized List<LineageInfo> getLineageInfoList() {
+  public synchronized List<LineageInfo> getLineageInfoList() throws LineageDoesNotExistException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/LineageMaster.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/LineageMaster.java [hunk 4]
```diff
@@ -441,7 +440,7 @@ public synchronized void queueForCheckpoint(CheckpointPlan plan) {
    * @throws InvalidPathException if the path is invalid
    */
   private synchronized List<CheckpointFile> pollToCheckpoint(long workerId)
-      throws FileDoesNotExistException, InvalidPathException {
+      throws FileDoesNotExistException, InvalidPathException, LineageDoesNotExistException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/LineageMaster.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/LineageMaster.java [hunk 5]
```diff
@@ -477,7 +476,8 @@ public synchronized void reportListFile(String path) throws FileDoesNotExistExce
    *
    * @param fileIds the id of the files
    */
-  public synchronized void requestFilePersistence(List<Long> fileIds) {
+  public synchronized void requestFilePersistence(List<Long> fileIds)
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/LineageMaster.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/LineageMaster.java [hunk 6]
```diff
@@ -504,7 +509,8 @@ private synchronized void requestFilePersistenceFromEntry(PersistFilesRequestEnt
    * @param workerId the worker id
    * @param persistedFiles the persisted files
    */
-  private synchronized void persistFiles(long workerId, List<Long> persistedFiles) {
+  private synchronized void persistFiles(long workerId, List<Long> persistedFiles)
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/LineageMaster.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 7]
```diff
@@ -119,14 +118,15 @@ public synchronized void completeFile(long fileId) {
    *
    * @param lineageId the lineage id
    */
-  public synchronized void deleteLineage(long lineageId) {
-    Preconditions.checkState(mIdIndex.containsKey(lineageId),
-        PreconditionMessage.LINEAGE_DOES_NOT_EXIST, lineageId);
+  public synchronized void deleteLineage(long lineageId) throws LineageDoesNotExistException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 8]
```diff
@@ -151,9 +151,9 @@ private void deleteLineage(long lineageId, Set<Long> deleted) {
    *
    * @param fileId the file id
    */
-  public synchronized void requestFilePersistence(long fileId) {
-    Preconditions.checkState(mOutputFileIndex.containsKey(fileId),
-        PreconditionMessage.LINEAGE_NO_OUTPUT_FILE, fileId);
+  public synchronized void requestFilePersistence(long fileId) throws LineageDoesNotExistException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 9]
```diff
@@ -174,9 +174,10 @@ public synchronized Lineage getLineage(long lineageId) {
    * @param lineage the lineage
    * @return the lineage's children
    */
-  public synchronized List<Lineage> getChildren(Lineage lineage) {
-    Preconditions.checkState(mIdIndex.containsKey(lineage.getId()),
-        PreconditionMessage.LINEAGE_DOES_NOT_EXIST, lineage.getId());
+  public synchronized List<Lineage> getChildren(Lineage lineage)
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: Javadoc

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 10]
```diff
@@ -187,9 +188,10 @@ public synchronized Lineage getLineage(long lineageId) {
    * @param lineage the lineage
    * @return the lineage's parents
    */
-  public synchronized List<Lineage> getParents(Lineage lineage) {
-    Preconditions.checkState(mIdIndex.containsKey(lineage.getId()),
-        "lineage id " + lineage.getId() + " does not exist");
+  public synchronized List<Lineage> getParents(Lineage lineage)
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: Javadoc

**Comment by yupeng9** [comment]
On top of this change, if you maintain a map from exception message to exception class, then you can still have a `TachyonPrecondition` like follows:

```
TachyonPrecondition.check(condition, message, args)
``

And `check` can be implemented like:
```
public void static check(boolean condition, ExceptionMessage message, Object ... args) {
     if(!condition) {
        message.getExceptionClassInstance().check();
     }
}
```

But you need to `check` not an instance method, and define it on the interface of TachyonException. What do you think?

**Comment by miasantreble** [comment]
@yupeng9 javadoc updated, I agree that having a single entry point(`TachyonPreconditions#check`) for all exceptions is more desirable, but might go beyond the scope of this PR. How about if I open another PR to address that?

**Comment by yupeng9** [comment]
LGTM. Thanks for the improvement!

**Comment by jsimsa** [comment]
LGTM

**Pull request merged**